PythonVisualizations/SortingBase.py

```diff
@@ -0,0 +1,420 @@
+            (len(self.list)+1) * self.CELL_SIZE):
+            # change cell width to 20
+            self.CELL_WIDTH = 20
+            # if this is the first insert that will change the width 
+            if not self.changeSize:
+                self.changeSize = True 
+                # redraw the existing cells to be narrower
+                self.fixCells()
+                self.redrawArrayCells()
+                # delete the cell values 
+                self.canvas.delete("cellVal")
+            
+        # If array needs to grow, add cells:
+        while self.size < len(self.list) + 1:
+            self.size += 1
+            self.createArrayCell(len(self.list))
+
+        # draw an index pointing to the last cell
+        indexDisplay = self.createIndex(len(self.list), 'nItems')
+        callEnviron |= set(indexDisplay)
+
+        # create new cell and cell value display objects
+        toPositions = (self.fillCoords(val, self.cellCoords(len(self.list))),
+                       self.cellCenter(len(self.list)))
+
+        # Animate arrival of new value from operations panel area
+        canvasDimensions = self.widgetDimensions(self.canvas)
+        startPosition = add_vector(
+            [canvasDimensions[0] // 2 - self.CELL_WIDTH, canvasDimensions[1]] * 2,
+            (0, 0) + (self.CELL_WIDTH - self.CELL_BORDER, 2.5*self.CELL_SIZE - self.CELL_BORDER))
+        
+        cellPair = self.createCellValue(startPosition, val)
+        callEnviron |= set(cellPair)
+        # if it is the smaller size cell
+        if self.changeSize:
+            # remove the generated cell value 
+            self.canvas.delete(cellPair[1])
+            callEnviron.remove(cellPair[1])
+            self.moveItemsTo((cellPair[0],), toPositions, steps=self.CELL_SIZE, sleepTime=0.01)
+        else:
+            self.moveItemsTo(cellPair, toPositions, steps=self.CELL_SIZE, sleepTime=0.01)
+
+        # add a new Drawable with the new value, color, and display objects
+        self.list.append(drawable(
+            val, self.canvas.itemconfigure(cellPair[0], 'fill'), *cellPair))
+        callEnviron ^= set(cellPair) # Remove new cell from temp call environment
+
+        # advance index for next insert
+        self.moveItemsBy(indexDisplay, (self.CELL_WIDTH, 0), sleepTime=0.01)
+
+        self.cleanUp(callEnviron)     
+       
+    def removeFromEnd(self):
+        if len(self.list) == 0:
+            self.setMessage('Array is empty!')
+            return
+    
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+        
+        # pop an Element from the list
+        n = self.list.pop()
+        
+        # Slide value rectangle up and off screen
+        if not self.changeSize:
+            items = (n.display_shape, n.display_val)
+        else:
+            items = (n.display_shape, )
+        callEnviron |= set(items)
+        self.moveItemsOffCanvas(items, N, sleepTime=0.02)
```

Comment:
Lines 196-199 would be simplified to `items = (n.display_shape, n.display_val)` if we make the modifications to allow some items to be `None` and ignored by routines such as `moveItemsOffCanvas()`.

Also the `nItems` index needs to be updated here.

Reply:
I changed this to move the `nItems` index but did not make the `moveItems` routines ignore `None`